### PR TITLE
[WIP] add fontconfig support to libass

### PR DESCRIPTION
**Pull request by copyliu**
add fontconfig support to libass, let it can read fonts from `fontdir`
and render subtitle with more controllable and flexible ass effects 

confirmed it works on Windows, but I can't get MacOS and iOS build success for now

### CODE REVIEW

**Comment by vit9696** [comment]
fontconfig support was removed from onscripter-ru in Ume branch in 2016 due to causing lags during subtitle initialisation on Windows. I do not think this got resolved at any time. Are you sure this is working properly now?

If it does

- the patch will need to build win32 version properly on Linux
- macOS/iOS builds should build fine
- the code needs to be reformatted (currently there are lots of whitespace and ending newline issues)
- gperf dependency should be removed, as it is not justified

I attached the relevant changesets: [r1202.patch.zip](https://github.com/umineko-project/onscripter-ru/files/4516379/r1202.patch.zip) and [r2364.patch.zip](https://github.com/umineko-project/onscripter-ru/files/4516403/r2364.patch.zip). Perhaps they can help.

File: .travis.yml [hunk 1]
```diff
@@ -23,6 +23,7 @@ matrix:
       - unset CC ; unset CXX ; unset CC_FOR_BUILD ; unset CXX_FOR_BUILD
       # homebrew addon is just broken
       - HOMEBREW_NO_AUTO_UPDATE=1 brew install yasm
+      - HOMEBREW_NO_AUTO_UPDATE=1 brew uninstall json-c
```
Comment (vit9696) [inline .travis.yml:26]: Why is this needed?

File: .travis.yml [hunk 2]
```diff
@@ -57,7 +58,7 @@ matrix:
       # Disable compiler overrides as they cause problems
       - unset CC ; unset CXX ; unset CC_FOR_BUILD ; unset CXX_FOR_BUILD
       # homebrew addon is just broken
-      - HOMEBREW_NO_AUTO_UPDATE=1 brew install yasm
+      - HOMEBREW_NO_AUTO_UPDATE=1 brew install yasm json-c
```
Comment (vit9696) [inline .travis.yml:61]: Why is this needed?

File: .travis.yml [hunk 3]
```diff
@@ -92,7 +93,7 @@ matrix:
         - libgl1-mesa-dev
         - libegl1-mesa-dev
         - libxrandr-dev
-
+        - gperf
```
Comment (vit9696) [inline .travis.yml:96]: I do not think it is needed.

File: .travis.yml [hunk 4]
```diff
@@ -111,45 +112,54 @@ matrix:
       - zip -qry onscripter-ru_linux64_r${BUILD}.zip onscripter-ru || exit 1
       - rm onscripter-ru ; cd ..
 
-  - os: linux
+  - os: windows
```
Comment (vit9696) [inline .travis.yml:115]: I am fine with fixing hosted build on Windows, but CI must build on Linux, as this is our primary test platform for cross builds.

File: .travis.yml [hunk 5]
```diff
@@ -162,6 +172,7 @@ matrix:
         - libgl1-mesa-dev
         - libegl1-mesa-dev
         - libxrandr-dev
+        - gperf
```
Comment (vit9696) [inline .travis.yml:175]: I do not think it is needed.

File: .travis.yml [hunk 6]
```diff
@@ -178,7 +189,6 @@ matrix:
       - mv DerivedData/Droid-package/onscripter-ru.apk ciout ; cd ciout
       - zip -qry onscripter-ru_droid_r${BUILD}.zip onscripter-ru.apk || exit 1
       - rm onscripter-ru.apk ; cd ..
-
```
Comment (vit9696) [inline .travis.yml:181]: Please avoid unnecessary whitespace changes.

File: Dependencies/build.sh [hunk 7]
```diff
@@ -402,30 +402,49 @@ extract_sources() {
...
-                        continue;;
-                esac;;
-        esac
+                case "$filetype" in
+                *application/x-tar*)
+                    cmd="tar -xf $file";;
+                *application/x-zip*|*application/zip*)
+                    cmd="unzip -q $file";;
+                *)
+                    # MinGW32 has broken mime types in 'file' command, fall back on the
```
Comment (vit9696) [inline Dependencies/build.sh:429]: I do not think this comment is valid here. Better move it above.

File: Dependencies/build.sh [hunk 8]
```diff
@@ -402,30 +402,49 @@ extract_sources() {
...
         local cmd=''
         local cmd_flags=''
-
-        case "$filetype" in
-            *application/x-tar*)
-                cmd='tar -xf';;
-            *application/x-zip*|*application/zip*)
-                cmd='unzip -q';;
+        case $(getTarget) in
+            win32)
```
Comment (vit9696) [inline Dependencies/build.sh:406]: This is not correct, as this is `getHost` that is broken, not target. I do not need these hacks on macOS.

File: Dependencies/build.sh [hunk 9]
```diff
@@ -402,30 +402,49 @@ extract_sources() {
...
-                        cmd='tar -xf';;
-                    zip)
-                        # should look one more level in, to see if tar is there...
-                        cmd='unzip -q';;
-                    *)
-                        continue;;
-                esac;;
-        esac
+                case "$filetype" in
+                *application/x-tar*)
```
Comment (vit9696) [inline Dependencies/build.sh:424]: Indentation got borked here.

File: Dependencies/pkgs/expat.pkgbuild [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+pkgrel=1
+depends=()
+sources=("https://github.com/libexpat/libexpat/releases/download/R_2_2_9/expat-2.2.9.tar.gz")
+hashes=('4456e0aa72ecc7e1d4b3368cd545a5eec7f9de5133a8dc37fdb1efa6174c4947')
+
+configopts=(
+    "--prefix=$outdir"
+    "--libdir=$outdir/lib"
+     '--disable-shared'
+)
```
Comment (vit9696) [inline Dependencies/pkgs/expat.pkgbuild:12]: Indentation is broken and missing newline.

File: Dependencies/pkgs/fontconfig.pkgbuild [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+     "--with-expat-lib=$outdir/lib"
+)
+prebuild() {
+    pushd "$pkgname-$pkgver" &>/dev/null
+    export FREETYPE_LIBS="-L$outdir/lib  -lfreetype -lharfbuzz"
+    export FREETYPE_CFLAGS="-I$outdir/include/freetype2"
+
+}
+postbuild() {
+    unset FREETYPE_LIBS
```
Comment (vit9696) [inline Dependencies/pkgs/fontconfig.pkgbuild:26]: Why do not you unset FREETYPE_CFLAGS? Please also fix the indentation and empty newlines within the funcitons and between htem.

File: Engine/Core/ONScripter.hpp [hunk 12]
```diff
@@ -1178,6 +1178,7 @@ class ONScripter : public ScriptParser {
 	int getCharacterPostDisplayDelay(char16_t codepoint, int speed);
 	int unpackInlineCall(const char *cmd, int &val);
 	const char *getFontPath(int i, bool fallback = true);
+    const char *getFontDir();
```
Comment (vit9696) [inline Engine/Core/ONScripter.hpp:1181]: Indentation got borked.

File: Engine/Core/Text.cpp [hunk 13]
```diff
@@ -1008,6 +1008,11 @@ const char *ONScripter::getFontPath(int i, bool /*fallback*/) {
 	return path;
 }
 
+const char *ONScripter::getFontDir() {
+    const char *path = fonts.fontdir;
+    return path;
```
Comment (vit9696) [inline Engine/Core/Text.cpp:1013]: Why not just `return fonts.fontdir;`?

**Comment by copyliu** [comment]
the early version of fontconfig may slow in Windows when it generate cache at first time, 
that was resolved in 2017 https://bugs.freedesktop.org/show_bug.cgi?id=64766 ,

but newer fontconfig version can't cross build for Windows in bionic due mingw's bug https://sourceforge.net/p/mingw-w64/mailman/message/35188208/ , https://github.com/rdp/ffmpeg-windows-build-helpers/issues/137
(but it builds in Debian 10)

gperf is need as fontconfig build dependency

Although I don't have a macOS environment I will try it in travis

File: Engine/Media/SubtitleDriver.cpp [hunk 14]
```diff
@@ -225,9 +225,9 @@ bool SubtitleDriver::init(int width, int height, const char *ass_sub_file, BaseR
 			ass_library = nullptr;
 			sendToLog(LogLevel::Error, "ass_renderer_init failed!\n");
 		} else {
-			//ass_set_fonts_dir(ass_library,nullptr);
+			ass_set_fonts_dir(ass_library, ons.getFontDir());
 			ass_set_frame_size(ass_renderer, width, height);
-			ass_set_fonts(ass_renderer, ons.getFontPath(currentFontID, true), "Sans", ASS_FONTPROVIDER_NONE, nullptr, 0);
+			ass_set_fonts(ass_renderer, ons.getFontPath(currentFontID, true), "Sans", ASS_FONTPROVIDER_AUTODETECT, nullptr, 1);
```
Comment (vit9696) [inline Engine/Media/SubtitleDriver.cpp:230]: What does 1 do here? From libass implementation it looks unused.

**Review by vit9696 — COMMENTED** [review]
Added some comments inline as well.

File: .travis.yml (hunk 1 above)
Comment (copyliu) [reply .travis.yml:26]: fontconfig build may fail in macOS with that json-c version, and it seems no configure options to disable it, just uninstall it that can pass as well

File: .travis.yml (hunk 2 above)
Comment (copyliu) [reply .travis.yml:61]: from some failed try before ;) , I will fix it later

File: .travis.yml (hunk 5 above)
Comment (copyliu) [reply .travis.yml:175]: fontconfig build dependency in Linux and Windows

**Comment by vit9696** [comment]
@copyliu 

I am reading the fontconfig bug, and it actually seems to me that prior the fixes landed libass ended up using native font libraries: https://bugs.freedesktop.org/show_bug.cgi?id=64766#c41. That, however, has nothing to do to us really.

Our use-case is to load embedded fonts from a particular directory (fonts). For this we do not need file system discovery or alike. libass has `ass_embedded_fonts_add_provider`, which can load actually fonts from a directory via `load_fonts_from_dir` function. To make the use of it we need to utilise the `ass_set_fonts_dir` call.

This makes me think there is no need in adding fontconfig dependency actually, we should make sure to pass a correct directory with the font files for this to work.

File: .travis.yml (hunk 1 above)
Comment (vit9696) [reply .travis.yml:26]: In general we write patches to disable such stuff, because if we need a library dependency, it needs to be statically linked into the resulting engine for this engine to work on consumer systems.

But in any case, I believe fontconfig is unnecessary. Please see my answer below.

**Comment by vit9696** [comment]
Now that I mentioned it, I am pretty sure we used `ass_set_fonts_dir` in the past. The reason we stopped using it was because it did not support Unicode on Windows properly. There is a patch here: https://github.com/umineko-project/onscripter-ru/blob/master/Dependencies/patches/libass-unicode-font-path-win.patch, and it will need to be extended for `ass_set_fonts_dir`.

**Comment by copyliu** [comment]
oh, I am not a  C/C++ programmer but `ass_embedded_fonts_add_provider` seems to be what I want, I will try to set it up without fontconfig.
the use case is , the Chinese subtitle need two or more fonts to display correctly, especially Chinese and Japanese bilingual subtitles.

**Comment by copyliu** [comment]
closed due to we don't need fontconfig at all
